design/StableRegions.md

```diff
@@ -0,0 +1,298 @@
+
+### region table
+
+ - purpose:
+   - relate a region ID to its size in pages.
+
+   - NB: This table exists as a "stable backup" of the per-region fields
+     that exist in the set of region objects, except for the access vectors' elements (see `block-region` table for their data).
+
+ - 32768 entries (statically sized).
+ - 8 bytes per entry.
+ - entry type = `RegionBlocks { size_in_pages: Nat64; }`
+ - the location of each entry in the table gives its corresponding region ID.
+ - the `size_in_pages` field measures the size of the region in terms of _pages_, not blocks.
+ - to do: track bit per region of whether in-use or reclaimed-and-available, somehow.
+
+
+### Overview of `rebuild`
+
+When upgrades work as expected, stable `Regions` are serialized and deserialized just like other stable data.
+
+For disaster recovery, we can **also** rebuild all of the region objects from data in stable memory.
+
+We use the `block-region` and `region` tables in stable memory to rebuild the regions' objects:
+
+ - The `region` table gives a size for each region, saying how large each vector for each region needs to be.
+ - The `block-region` table gives a relative position and region ID for each block ID.
+
+Once each regions' vectors have been allocated, the block-region table says how to fill them, one entry at a time.
+Unlike the Rust design, vector entries can be populated out-of-order.
+
+
+### Special (reserved) regions
+
+  - Region 0 -- Anonymous region, for supporting the legacy API that we have today, which lacks `Region` values.
+  - Region 1 -- "Reclaimed blocks" region that consists of blocks reclaimed from GC'd regions.
+  - Regions 2-15 -- Future use by Motoko RTS (TBD).
+
+### Overview of GC support (future work)
+
+- Regions are reprensented (see special subsection) with heap objects that are `stable`, but mutable.
+- They have special GC headers to recognize their special structure.
+- The Region table keeps track of which available IDs are in use as Region heap values are GC'd.
+
+#### Region allocation, re-allocation
+
+A region with a particular ID is first initially allocated by a global, bumped counter reaching it.
+
+When that happens, the region table is marked (TO DO) indicating that region is in use.
+
+The region heap object with that ID witnesses its lifetime, and is reclaimed by GC when it goes out of scope.
+
+When the bump counter reaches its maximum value, the allocation algorithm changes to scanning the regions,
+looking for the first available one (with minimum ID).
+
+#### Region reclamation
+
+A heap object with a certain region ID lives until GC reclaims it, where the region table is re-marked (TO DO) to indicate that the region is free again.
+
+Before reusing that region, each of its block are added to the special "free blocks" region.
+
+
+### Migration from earlier designs into region system
+
+#### Stable memory format versions
+
+The first 32 bits of stable memory record a version number.
```

Comment:
I don't think that's quite true: From design/StableMemory.md, the layout depends on the logical size of experimental stable memory, I think.

Stable memory layout (between upgrades), assuming stable variable encoding `v == {fs:vs}`, a record value of record type {fs:Ts}.

NOTE: A program with no stable variables still writes an empty record value `v = {}`.

```
(case !size == 0) // hence N = 0
  [0..3] StableVariable data len
  [4..4+len-1] StableVariable data
  [4+len-1,..M-1] 0...0 // zero padding
(case !size > 0)
[0..3]  0...0
[4..N-1]  StableMemory bytes
[N..N+3]  StableVariable data len
[N+4..(N+4)+len-1] StableVariable data
[(N+4)+len..M-13] 0...0 // zero padding
[M-12..M-9] value N/64Ki = !size
[M-8..M-5] saved StableMemory bytes
[M-4..M-1]  version word

where N = !size * pagesize // logical memory size
      M = ic0.stable_size() * pagesize // physical memory size
      pagesize = 64Kb (2^16 bytes)
where (len, data) = serialize<Ts>(v,data)

```
If the actor just used stable variables but never allocated stable memory pages using the ExperimentalStableMemory library (ESM), then word 0 is non-zero and contains the length of following stable variables data (the length of serialized data can never be 0, even if the program has no stable variables).
Otherwise, the program was using ESM and wrote marker 0 to word [0...3], saving the contents at [M-8..M-5], and left ESM pages otherwise unchanged, and wrote stable variables at address [N to N+3] (stable variable size) and [N+4, N+4+len-1] and a version number and other metadata to the last address in physical stable memory (allocating another page if necessary if stable variable data didn't leave enough space). So the actual version number (currently 1), if any, is in the least word of the last page.
I think you want your new scheme to use 2 for the new version, and migrate data only if the version read is 1.